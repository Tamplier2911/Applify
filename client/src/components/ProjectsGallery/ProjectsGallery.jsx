// import "./ProjectsGallery.scss";
import React, { useState, useEffect } from "react";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentImage } from "../../redux/projects/projects.selectors";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import {
  setCurrentImageNext,
  setCurrentImagePrev,
} from "../../redux/projects/projects.actions";

// data transformations
import getImageRelativePath from "../../utils/PathTransformations/getImageRelativePath";

// JS Rendering CSS
import {
  ProjectsGalleryContainer,
  ProjectsGalleryTitle,
  ProjectsGalleryComposition,
  ProjectsGalleryArrow,
  ProjectsGalleryArrowLeftSVG,
  ProjectsGalleryArrowRightSVG,
  ProjectsGalleryImageContainer,
  ProjectsGalleryImage,
} from "./ProjectsGalleryStyles";

// component constants
import galleryData from "./ProjectsGalleryConstants";

// get window dimensions
import getWindowDimensions from "../../utils/WindowCaptions/getWindowDimensions";
// get device value
import getDeviceValue from "../../utils/WindowCaptions/getDeviceValue";
// debouncer
import { debounce } from "../../utils/Debounce/debounce";

const ProjectsGallery = ({
  currentImage,
  setCurrentImageNext,
  setCurrentImagePrev,
  lang,
}) => {
  const [isLoaded, setIsLoaded] = useState(false);
  const [windowDimensions, setWindowDimensions] = useState(
    getWindowDimensions()
  );
  const { width } = windowDimensions;

  useEffect(() => {
    setIsLoaded(false);
    setTimeout(() => {
      setIsLoaded(true);
    }, 600);
    // 400

    // handle resize
    const debounced = debounce(
      () => setWindowDimensions(getWindowDimensions()),
      250
    );
    window.addEventListener("resize", debounced);

    // const handleResize = () => setWindowDimensions(getWindowDimensions());
    // window.addEventListener("resize", handleResize);
    // return () => window.removeEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", debounced);
  }, [currentImage]);

  const { projectsGalleryHeader } = galleryData[lang];
  const { image } = currentImage;

  // define image path
  const projectPortrait = getImageRelativePath(image || "");
  // get divice value
  const device = getDeviceValue(width);

  return (
    <ProjectsGalleryContainer animateIn={"bounceInLeft"} animateOnce={true}>
      <ProjectsGalleryTitle>{projectsGalleryHeader}</ProjectsGalleryTitle>
      <ProjectsGalleryComposition>
        <ProjectsGalleryArrow onClick={() => setCurrentImagePrev(device)}>
          <ProjectsGalleryArrowLeftSVG />
        </ProjectsGalleryArrow>

        <ProjectsGalleryImageContainer load={isLoaded}>
          <ProjectsGalleryImage
            src={projectPortrait}
            alt="project gallery overview"
          ></ProjectsGalleryImage>
        </ProjectsGalleryImageContainer>

        <ProjectsGalleryArrow onClick={() => setCurrentImageNext(device)}>
          <ProjectsGalleryArrowRightSVG />
        </ProjectsGalleryArrow>
      </ProjectsGalleryComposition>
    </ProjectsGalleryContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  currentImage: selectCurrentImage,
  lang: selectCurrentLanguage,
});

export default connect(mapStateToProps, {
  setCurrentImageNext,
  setCurrentImagePrev,
})(ProjectsGallery);
