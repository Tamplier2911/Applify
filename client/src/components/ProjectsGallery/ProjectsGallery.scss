/*
.projectsGallery {
  display: grid;
  // grid-template-columns: repeat(10, 1fr);
  // grid-template-rows: repeat(10, min-content);
  // padding: 1rem;

  &__header {
    // grid-area: 1 / 1 / 2 / -1;
    font-family: var(--font-title);
    font-size: 4rem;
    font-weight: 300;
    padding: 1rem;
  }

  &__gallery {
    display: grid;
    grid-template-columns: min-content 1fr min-content;
    grid-column-gap: 2rem;
  }

  &__arrow-left {
    align-self: center;
  }

  &__arrow-right {
    align-self: center;
  }

  &__arrow-svg {
    cursor: pointer;
    fill: var(--cl-tiles-2);
    width: 5rem;
    height: 5rem;
    transition: fill 0.3s;

    background-color: rgb(243, 243, 243);
    padding: 1rem;
    border-radius: 50%;
    box-shadow: 0 0.2rem 1rem var(--cl-square-shadow);

    &:hover {
      fill: var(--cl-tiles-1);
      animation: pulse 2s infinite;

      @media (hover: none) {
        animation: none;
      }
    }

    @media only screen and (max-width: 720px) {
      width: 4rem;
      height: 4rem;
    }

    @media only screen and (max-width: 425px) {
      width: 3.5rem;
      height: 3.5rem;
    }
  }

  &__imgContainer {
    box-shadow: 0 0.2rem 1rem var(--cl-square-shadow);
  }

  &__img {
    width: 100%;
    height: 100%;
    display: block;
    object-fit: cover;
  }
}

@-webkit-keyframes pulse {
  0% {
    -webkit-box-shadow: 0 0 0 0 rgba(0, 168, 168, 0.4);
  }
  70% {
    -webkit-box-shadow: 0 0 0 15px rgba(204, 169, 44, 0);
  }
  100% {
    -webkit-box-shadow: 0 0 0 0 rgba(204, 169, 44, 0);
  }
}
@keyframes pulse {
  0% {
    -moz-box-shadow: 0 0 0 0 rgba(0, 168, 168, 0.4);
    box-shadow: 0 0 0 0 rgba(0, 168, 168, 0.4);
  }
  70% {
    -moz-box-shadow: 0 0 0 15px rgba(204, 169, 44, 0);
    box-shadow: 0 0 0 15px rgba(204, 169, 44, 0);
  }
  100% {
    -moz-box-shadow: 0 0 0 0 rgba(204, 169, 44, 0);
    box-shadow: 0 0 0 0 rgba(204, 169, 44, 0);
  }
}

*/
