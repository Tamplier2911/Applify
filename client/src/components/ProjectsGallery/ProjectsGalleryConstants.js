const galleryData = {
  eng: {
    projectsGalleryHeader: "Projects gallery:"
  },
  rus: { projectsGalleryHeader: "Галерея проектов:" },
  ukr: { projectsGalleryHeader: "Галерея проектів:" }
};

export default galleryData;
