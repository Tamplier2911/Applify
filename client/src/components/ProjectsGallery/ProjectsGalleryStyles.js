import styled, { css } from "styled-components";

import { ReactComponent as ChevronLeftSVG } from "../../assets/svg/chevron-left.svg";
import { ReactComponent as ChevronRightSVG } from "../../assets/svg/chevron-right.svg";

const fadeInImage = props => {
  const { load } = props;
  if (!load) {
    return `opacity: 0;`;
  } else if (load) {
    return `transition: opacity 1s ease-in;
    opacity: 1;`;
  }
};

/*
const ImageAnimation = css`
  @keyframes imageFadein {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
`;
*/

const ArrowStyles = css`
  align-self: center;
`;

const ArrowSVGStyles = css`
  cursor: pointer;
  fill: var(--cl-tiles-2);
  width: 5rem;
  height: 5rem;
  transition: fill 0.3s, background-color 0.3s;

  background-color: var(--cl-projects-gal-bg-col);
  padding: 1rem;
  border-radius: 50%;
  box-shadow: 0 0.2rem 1rem var(--cl-square-shadow);

  &:hover {
    fill: var(--cl-mainhover);
    background-color: var(--white);
    animation: pulse 2s infinite;

    @media (hover: none) {
      animation: none;
    }
  }

  @media only screen and (max-width: 720px) {
    width: 4rem;
    height: 4rem;
  }

  @media only screen and (max-width: 425px) {
    width: 3.5rem;
    height: 3.5rem;
  }
`;

export const ArrowsAnimation = css`
  @-webkit-keyframes pulse {
    0% {
      -webkit-box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse);
    }
    70% {
      -webkit-box-shadow: 0 0 0 15px var(--cl-projects-gal-pulse-shadow);
    }
    100% {
      -webkit-box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse-shadow);
    }
  }
  @keyframes pulse {
    0% {
      -moz-box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse);
      box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse);
    }
    70% {
      -moz-box-shadow: 0 0 0 15px var(--cl-projects-gal-pulse-shadow);
      box-shadow: 0 0 0 15px var(--cl-projects-gal-pulse-shadow);
    }
    100% {
      -moz-box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse-shadow);
      box-shadow: 0 0 0 0 var(--cl-projects-gal-pulse-shadow);
    }
  }
`;

export const ProjectsGalleryContainer = styled.div`
  display: grid;
`;

export const ProjectsGalleryTitle = styled.h2`
  font-family: var(--font-title);
  font-size: 4rem;
  font-weight: 300;
  padding: 1rem;
`;

export const ProjectsGalleryComposition = styled.div`
  display: grid;
  grid-template-columns: min-content 1fr min-content;
  grid-column-gap: 2rem;
`;

export const ProjectsGalleryArrow = styled.div`
  ${ArrowStyles}
`;

export const ProjectsGalleryArrowLeftSVG = styled(ChevronLeftSVG)`
  ${ArrowSVGStyles}
  ${ArrowsAnimation}
`;

export const ProjectsGalleryArrowRightSVG = styled(ChevronRightSVG)`
  ${ArrowSVGStyles}
  ${ArrowsAnimation}
`;

export const ProjectsGalleryImageContainer = styled.div`
  ${fadeInImage}
  box-shadow: 0 0.2rem 1rem var(--cl-square-shadow);
`;

export const ProjectsGalleryImage = styled.img`
  width: 100%;
  height: 100%;
  display: block;
  object-fit: cover;
`;
