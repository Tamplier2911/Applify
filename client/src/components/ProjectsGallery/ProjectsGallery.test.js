import React from "react";
import toJson from "enzyme-to-json";

// projects
import projectsData from "../../redux/projects/projectsConstants";

// component
import ProjectsGallery from "./ProjectsGallery";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
} from "../../tests/testUtils";

describe("<ProjectsGallery />", () => {
  const initialState = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
    projects: projectsData,
  };
  const store = storeFactory(initialState);

  let wrapper;
  beforeEach(() => {
    wrapper = setShallowC(<ProjectsGallery store={store} />);
  });

  it("renders without an error", () => {
    const projectsGallery = findByTestAttr(wrapper, "projects-gallery");
    expect(projectsGallery).toHaveLength(1);
  });

  it("renders children - project image", () => {
    const projectsGalleryImage = findByTestAttr(
      wrapper,
      "projects-gallery-image"
    );
    expect(projectsGalleryImage).toHaveLength(1);
  });

  it("selects next image on setCurrentImageNext", () => {
    const projectsGalleryNextBtn = findByTestAttr(
      wrapper,
      "projects-gallery-next-btn"
    );
    // ensure button mounted
    expect(projectsGalleryNextBtn).toHaveLength(1);
    // grab current state
    const oldState = store.getState();
    // assert to have expected state
    expect(oldState.projects.currentImage.id).toBe(100);
    // simulate btn click
    projectsGalleryNextBtn.simulate("click");
    // grab new state
    const newState = store.getState();
    // assert new state to be expected state
    expect(newState.projects.currentImage.id).toBe(101);
  });

  it("selects previous image on setCurrentImagePrev", () => {
    const projectsGalleryPrevBtn = findByTestAttr(
      wrapper,
      "projects-gallery-prev-btn"
    );
    // ensure btn mounted
    expect(projectsGalleryPrevBtn).toHaveLength(1);
    // grab current state
    const oldState = store.getState();
    // assert to have expected state
    expect(oldState.projects.currentImage.id).toBe(101);
    // simulate btn click
    projectsGalleryPrevBtn.simulate("click");
    // get new state
    const newState = store.getState();
    // assert new state to be expected
    expect(newState.projects.currentImage.id).toBe(100);
  });

  it("matches snapshot", () => {
    const projectsGallery = findByTestAttr(wrapper, "projects-gallery");
    expect(toJson(projectsGallery)).toMatchSnapshot();
  });
});
