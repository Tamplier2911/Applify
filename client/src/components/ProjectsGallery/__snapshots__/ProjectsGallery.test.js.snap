// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProjectsGallery /> matches snapshot 1`] = `
<Styled(ScrollAnimation)
  animateIn="bounceInLeft"
  animateOnce={true}
  data-test="projects-gallery"
>
  <styled.h2>
    Projects gallery:
  </styled.h2>
  <styled.div>
    <styled.div
      data-test="projects-gallery-prev-btn"
      onClick={[Function]}
    >
      <Styled(Component) />
    </styled.div>
    <styled.div
      load={false}
    >
      <styled.img
        alt="project gallery overview"
        data-test="projects-gallery-image"
        src="http://localhost/api/uploads/images/projects/audit-desktop.jpg"
      />
    </styled.div>
    <styled.div
      data-test="projects-gallery-next-btn"
      onClick={[Function]}
    >
      <Styled(Component) />
    </styled.div>
  </styled.div>
</Styled(ScrollAnimation)>
`;
